### PR TITLE
enhance(plonky2x): Upgrade test engine

### PR DESCRIPTION
**Pull request by Owliie**
Added:
- time measurement for build and tests
- ssz decoder
- file or folder tests
- wrapper for weigh justification and finalization circuit

### CODE REVIEW

File: casper-finality-proofs/src/test_engine/TestEngine.md [hunk 1]
```diff
@@ -50,7 +50,7 @@ The test engine is a tool for running unit tests for plonky2 circuits.
 
   To test a circuit, create a wrapper in `src/test_engine/wrappers/`. It represents a function that writes input data to the circuit and asserts its outputs. It uses `TestData` data to assert that the circuit is working correctly.
 
-  `path` is an argument to the `wrapper()` method, received from the test engine. It is the path to the JSON file containing the input and output data for the test.
+  `path` is an argument to the `wrapper()` method, received from the test engine. It is the path to the JSON or YAML file containing the input and output data for the test.
```
Comment (Xearty) [inline casper-finality-proofs/src/test_engine/TestEngine.md:53]: or directory

File: casper-finality-proofs/bin/weigh_justification_and_finalization.rs [hunk 2]
```diff
@@ -459,6 +182,5 @@ fn test_circuit_sample_data() {
 }
 
 fn main() {
-    test_circuit_ssz_snappy();
-    // test_circuit_sample_data();
+    test_circuit_sample_data();
```
Comment (Xearty) [inline casper-finality-proofs/bin/weigh_justification_and_finalization.rs:185]: Delete the test_circuit_sample_data function and dump its body in main

File: casper-finality-proofs/src/test_engine/TestEngine.md [hunk 3]
```diff
@@ -93,8 +95,13 @@ The test engine is a tool for running unit tests for plonky2 circuits.
...
      pub fn map_test_to_wrapper(test: TestWrappers) -> ... {
          match test {
-             TestWrappers::NewTestCircuit => Box::new(|path, should_assert| wrapper_new_test(path.as_str(), should_assert)),
-             ...
+              TestWrappers::NewTestCircuit => (
+                Box::new(|| {
+                  Lazy::force(&circuit_new_test);
+                }),
+                Box::new(|path, should_assert| wrapper_new_test(path.as_str(), should_assert)),
+              ),
```
Comment (Xearty) [inline casper-finality-proofs/src/test_engine/TestEngine.md:103]: Can you make this a macro?

File: casper-finality-proofs/src/test_engine/TestEngine.md [hunk 4]
```diff
@@ -93,8 +95,13 @@ The test engine is a tool for running unit tests for plonky2 circuits.
      ```rust
      pub fn map_test_to_wrapper(test: TestWrappers) -> ... {
          match test {
-             TestWrappers::NewTestCircuit => Box::new(|path, should_assert| wrapper_new_test(path.as_str(), should_assert)),
-             ...
+              TestWrappers::NewTestCircuit => (
+                Box::new(|| {
+                  Lazy::force(&circuit_new_test);
+                }),
+                Box::new(|path, should_assert| wrapper_new_test(path.as_str(), should_assert)),
```
Comment (Xearty) [inline casper-finality-proofs/src/test_engine/TestEngine.md:102]: Rename this to should_fail?

File: casper-finality-proofs/src/test_engine/utils/extract_merkle_info.rs
Comment (Xearty) [inline casper-finality-proofs/src/test_engine/utils/extract_merkle_info.rs:1]: Rename this to test_data_generation?

**Review by NikolayKostadinov21 — APPROVED** [review]

File: casper-finality-proofs/src/test_engine/utils/extract_merkle_info.rs
Comment (Owliie) [reply casper-finality-proofs/src/test_engine/utils/extract_merkle_info.rs:1]: I suppose if we have more data extraction/generation tools, we will dump them in other files, so we should name these helpers accordingly. Also, it is in the test engine directory therefore "test" is meant by default.

Comment (Xearty) [reply casper-finality-proofs/src/test_engine/utils/extract_merkle_info.rs:1]: Merkle info doesn't tell much. Plus, there are other things in there like balances extraction. That's why I suggested this change which I still believe makes sense.

Comment (Xearty) [reply casper-finality-proofs/src/test_engine/utils/extract_merkle_info.rs:1]: The name could be more customized but I think it should be something i that spirit

**Review by Xearty — APPROVED** [review]

**Pull request merged**